File: 01-Programming-Languages/PHP/Data-Structures/Queue/queue_using_stacks.php
<?php

/*
Problem: Implement Queue Using Stacks
Description: Implement a FIFO queue using two stacks (push, pop, peek, empty).

Approach:
- Use two stacks: one for input, one for output
- On pop/peek, if output is empty, transfer all elements from input to output
- This reverses order, making the oldest element accessible first

Time Complexity: O(1) amortized per operation
Space Complexity: O(n)

Example:
Input:  enqueue(1), enqueue(2), peek(), dequeue(), empty()
Output: 1, 1, false
*/

class MyQueue {
    private array $in;
    private array $out;

    public function __construct() {
        $this->in = [];
        $this->out = [];
    }

    public function enqueue($x): void {
        $this->in[] = $x;
    }

    private function transfer(): void {
        if (empty($this->out)) {
            while (!empty($this->in)) {
                $this->out[] = array_pop($this->in);
            }
        }
    }

    public function dequeue() {
        $this->transfer();
        return array_pop($this->out);
    }

    public function peek() {
        $this->transfer();
        return end($this->out);
    }

    public function empty(): bool {
        return empty($this->in) && empty($this->out);
    }
}

$q = new MyQueue();
$q->enqueue(1);
$q->enqueue(2);
echo "Peek: " . $q->peek() . "\n";
echo "Dequeue: " . $q->dequeue() . "\n";
echo "Empty: " . ($q->empty() ? "true" : "false") . "\n";
echo "Dequeue: " . $q->dequeue() . "\n";
echo "Empty: " . ($q->empty() ? "true" : "false") . "\n";

File: 01-Programming-Languages/Ruby/Data-Structures/Queue/queue_using_stacks.rb
=begin
Problem: Implement Queue Using Stacks
Description: Implement a FIFO queue using two stacks.

Approach:
- Use two stacks: one for input, one for output
- On pop/peek, if output is empty, transfer all from input to output

Time Complexity: O(1) amortized per operation
Space Complexity: O(n)

Example:
Input:  enqueue(1), enqueue(2), peek(), dequeue(), empty()
Output: 1, 1, false
=end

class MyQueue
  def initialize
    @in = []
    @out = []
  end

  def enqueue(x)
    @in.push(x)
  end

  def dequeue
    transfer
    @out.pop
  end

  def peek
    transfer
    @out.last
  end

  def empty?
    @in.empty? && @out.empty?
  end

  private

  def transfer
    if @out.empty?
      @out.push(@in.pop) until @in.empty?
    end
  end
end

q = MyQueue.new
q.enqueue(1)
q.enqueue(2)
puts "Peek: #{q.peek}"
puts "Dequeue: #{q.dequeue}"
puts "Empty: #{q.empty?}"
puts "Dequeue: #{q.dequeue}"
puts "Empty: #{q.empty?}"
